File: R-analysis/scripts/generate_sankey.R
# 加载library
library(networkD3)
library(dplyr)

# 读取Kraken2报告文件
kraken_report <- read.table("/ResultData/kraken_taxonomy_m11213.report", header = TRUE, sep = "\t")

# 检查kraken_report数据框的内容
print(head(kraken_report))
print(names(kraken_report))

# 假设rank列实际上是U列，name列实际上是unclassified列
kraken_report$rank <- as.character(kraken_report$U)
kraken_report$name <- as.character(kraken_report$unclassified)

# 过滤无效行（例如 rank 为 'U' 的行）
kraken_report <- kraken_report %>% filter(rank != "U")

# 创建节点列表
nodes <- data.frame(name = unique(c(kraken_report$rank, kraken_report$name)))

# 创建链接列表
links <- data.frame(
  source = match(kraken_report$rank, nodes$name) - 1,
  target = match(kraken_report$name, nodes$name) - 1,
  value = kraken_report$X0
)

# 生成Sankey图
sankey <- sankeyNetwork(
  Links = links,
  Nodes = nodes,
  Source = "source",
  Target = "target",
  Value = "value",
  NodeID = "name",
  units = "Reads",
  fontSize = 12,
  nodeWidth = 30
)

# 保存Sankey图为HTML文件
saveNetwork(sankey, "sankey_diagram.html", selfcontained = FALSE)

# 或者显示Sankey图
sankey

File: scripts/legacy_scripts/generate_sankey.py
import pandas as pd
import plotly.graph_objects as go

# 读取 Kraken2 报告文件
kraken_report_file = "/media/shuai/TOSHBA/mm-dev/ResultData/kraken_taxonomy_m11213.report"
kraken_report = pd.read_csv(kraken_report_file, sep='\t', header=None, names=["percent", "reads", "taxReads", "rank", "taxID", "name"])

# 过滤无效行
kraken_report = kraken_report[kraken_report["rank"] != "U"]

# 创建层级关系
def create_hierarchy(df):
    hierarchy = {}
    for i, row in df.iterrows():
        lineage = row["name"].split('|')
        parent = None
        for node in lineage:
            if parent:
                if parent not in hierarchy:
                    hierarchy[parent] = []
                if node not in hierarchy[parent]:
                    hierarchy[parent].append(node)
            parent = node
    return hierarchy

hierarchy = create_hierarchy(kraken_report)

# 生成节点和链接数据
nodes = []
links = []

def add_to_sankey(parent, children, source_idx):
    for child in children:
        if child not in nodes:
            nodes.append(child)
        target_idx = nodes.index(child)
        links.append({
            "source": source_idx,
            "target": target_idx,
            "value": kraken_report[kraken_report["name"] == child]["taxReads"].sum()
        })
        if child in hierarchy:
            add_to_sankey(child, hierarchy[child], target_idx)

# 根节点处理
root_nodes = kraken_report[kraken_report["rank"] == "D"]["name"].unique()
for root_node in root_nodes:
    if root_node not in nodes:
        nodes.append(root_node)
    root_idx = nodes.index(root_node)
    if root_node in hierarchy:
        add_to_sankey(root_node, hierarchy[root_node], root_idx)

# 生成 Sankey 图
fig = go.Figure(data=[go.Sankey(
    node=dict(
        pad=15,
        thickness=20,
        line=dict(color="black", width=0.5),
        label=nodes
    ),
    link=dict(
        source=[link["source"] for link in links],
        target=[link["target"] for link in links],
        value=[link["value"] for link in links]
    )
)])

# 更新图表布局
fig.update_layout(title_text="Kraken2 Sankey Diagram", font_size=10)

# 保存 Sankey 图为 HTML 文件
fig.write_html("sankey_diagram.html")

# 保存 Sankey 图为 PNG 文件
fig.write_image("sankey_diagram.png")
